blockchain/chain/src/store/chain_store.rs

```diff
@@ -646,6 +739,62 @@ where
     Ok(out)
 }
 
+#[cfg(feature = "json")]
+pub mod headchange_json {
+    use super::*;
+    use blocks::tipset_json::TipsetJsonRef;
+    use serde::Serialize;
+
+    #[derive(Debug, Serialize, Clone)]
+    #[serde(rename_all = "lowercase")]
+    #[serde(tag = "type", content = "val")]
+    pub enum HeadChangeJson<'a> {
+        Current(TipsetJsonRef<'a>),
+        Apply(TipsetJsonRef<'a>),
+        Revert(TipsetJsonRef<'a>),
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct IndexToHeadChangeJson(pub usize, pub HeadChange);
+    impl<'a> From<&'a HeadChange> for HeadChangeJson<'a> {
+        fn from(wrapper: &'a HeadChange) -> Self {
+            match wrapper {
+                HeadChange::Current(tipset) => HeadChangeJson::Current(TipsetJsonRef(&tipset)),
+                HeadChange::Apply(tipset) => HeadChangeJson::Apply(TipsetJsonRef(&tipset)),
+                HeadChange::Revert(tipset) => HeadChangeJson::Revert(TipsetJsonRef(&tipset)),
+            }
+        }
+    }
+
+    pub async fn sub_head_changes(
+        publisher: Arc<RwLock<Publisher<IndexToHeadChangeJson>>>,
```

Comment:
The publisher is needed here because you need to make sure you are iterating from the first the beginning of the channel in situations where RPCState is being shared amongst processes

Reply:
Maybe I'm confused about how this is being published to multiple threads, I can look into this later.